File: FindAllDuplicatesInAnArray.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int n = nums.size();
        vector<int> result;
        for(int i=0;i<n;i++){
            int j = abs(nums[i])-1;
            if(nums[j]<0){
                result.push_back(j+1);
            }
            else{
                nums[j] = -nums[j];
            }
        }
        return result;
    }
};

File: FindAllDuplicatesInAnArray.go
func findDuplicates(nums []int) []int {
    //we can make use of index of nums
    n := len(nums)
    res := []int{}
    for i := 0; i < n; i++ {
        k := nums[i]
        if nums[abs(k)-1] < 0 {
            res = append(res, abs(k))
        }
        nums[abs(k)-1] = -nums[abs(k)-1]
    }
    return res
}

func abs(x int) int {
    if x >= 0 {
        return x
    } else {
        return -x
    }
}
